<?xml version="1.0"?>
<robot name="puppet" xmlns:xacro="http://www.ros.org/wiki/xacro">
  
  <!-- Base Link (Root) -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.2 0.2 0.3"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
  </link>
  
  <!-- Head Joint -->
  <joint name="head_joint" type="revolute">
    <parent link="base_link"/>
    <child link="head"/>
    <origin xyz="0 0 0.15"/>
    <axis xyz="1 0 0"/>
    <limit effort="1.0" lower="-0.5" upper="0.5" velocity="1.0"/>
  </joint>
  
  <link name="head">
    <visual>
      <geometry>
        <sphere radius="0.1"/>
      </geometry>
      <material name="red">
        <color rgba="1 0 0 1"/>
      </material>
    </visual>
  </link>
  
  <!-- Left Arm -->
  <joint name="left_arm_joint" type="revolute">
    <parent link="base_link"/>
    <child link="left_arm"/>
    <origin xyz="-0.15 0 0.1"/>
    <axis xyz="0 1 0"/>
    <limit effort="1.0" lower="-1.0" upper="1.0" velocity="1.0"/>
  </joint>
  
  <link name="left_arm">
    <visual>
      <origin xyz="-0.075 0 0" rpy="0 1.5708 0"/>
      <geometry>
        <cylinder radius="0.03" length="0.15"/>
      </geometry>
      <material name="green">
        <color rgba="0 1 0 1"/>
      </material>
    </visual>
  </link>
  
  <!-- Right Arm -->
  <joint name="right_arm_joint" type="revolute">
    <parent link="base_link"/>
    <child link="right_arm"/>
    <origin xyz="0.15 0 0.1"/>
    <axis xyz="0 1 0"/>
    <limit effort="1.0" lower="-1.0" upper="1.0" velocity="1.0"/>
  </joint>
  
  <link name="right_arm">
    <visual>
      <origin xyz="0.075 0 0" rpy="0 1.5708 0"/>
      <geometry>
        <cylinder radius="0.03" length="0.15"/>
      </geometry>
      <material name="green">
        <color rgba="0 1 0 1"/>
      </material>
    </visual>
  </link>
  
  <!-- Left Leg -->
  <joint name="left_leg_joint" type="revolute">
    <parent link="base_link"/>
    <child link="left_leg"/>
    <origin xyz="-0.08 0 -0.15"/>
    <axis xyz="1 0 0"/>
    <limit effort="1.0" lower="-1.0" upper="1.0" velocity="1.0"/>
  </joint>
  
  <link name="left_leg">
    <visual>
      <origin xyz="0 0 -0.1" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.04" length="0.2"/>
      </geometry>
      <material name="yellow">
        <color rgba="1 1 0 1"/>
      </material>
    </visual>
  </link>
  
  <!-- Right Leg -->
  <joint name="right_leg_joint" type="revolute">
    <parent link="base_link"/>
    <child link="right_leg"/>
    <origin xyz="0.08 0 -0.15"/>
    <axis xyz="1 0 0"/>
    <limit effort="1.0" lower="-1.0" upper="1.0" velocity="1.0"/>
  </joint>
  
  <link name="right_leg">
    <visual>
      <origin xyz="0 0 -0.1" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.04" length="0.2"/>
      </geometry>
      <material name="yellow">
        <color rgba="1 1 0 1"/>
      </material>
    </visual>
  </link>
</robot>
